### PR TITLE
perf: represent pc:hint mapping as a vector indexed by ranges

### PR DESCRIPTION
**Pull request by Oppen**
Replaces the use of `HashMap<usize, _>` to store hint data with `Vec<_>` and
ranges in the form `Vec<Option<(usize, NonZeroUsize)>>`, where the index
corresponds to the given PC, the first field is the first index in the hints
`Vec` and the second one is the length of such range.  
This avoids the cost of hashing and reduces memory fragmentation by collapsing
all of the individual `Vec`s of each PC into a single bigger `Vec`. The cost of
this is two extra machine words per PC for storing the ranges, as well as some
increase in code complexity.

## Benchmark results

In files suffixed by `-proof` additional arguments `--memory_file /dev/null --trace_file /dev/null --proof_mode` are passed. `-quick` runs don't generate traces nor dump memory.

blake2s_integration_benchmark-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 6.175 ± 0.055 | 6.098 | 6.334 | 1.20 ± 0.01 |
| `base-bigint` | 5.686 ± 0.043 | 5.638 | 5.921 | 1.10 ± 0.01 |
| `ranges-felt` | 5.564 ± 0.024 | 5.516 | 5.612 | 1.08 ± 0.01 |
| `ranges-bigint` | 5.150 ± 0.029 | 5.127 | 5.278 | 1.00 |

blake2s_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 820.3 ± 6.8 | 810.9 | 846.5 | 1.08 ± 0.01 |
| `base-bigint` | 1004.3 ± 5.9 | 994.6 | 1023.1 | 1.32 ± 0.02 |
| `ranges-felt` | 761.2 ± 8.4 | 736.9 | 778.2 | 1.00 |
| `ranges-bigint` | 983.7 ± 14.2 | 939.9 | 1012.9 | 1.29 ± 0.02 |

compare_arrays_200000-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.750 ± 0.018 | 1.712 | 1.796 | 1.09 ± 0.02 |
| `base-bigint` | 1.972 ± 0.009 | 1.956 | 1.992 | 1.22 ± 0.02 |
| `ranges-felt` | 1.612 ± 0.019 | 1.581 | 1.688 | 1.00 |
| `ranges-bigint` | 1.805 ± 0.014 | 1.774 | 1.852 | 1.12 ± 0.02 |

compare_arrays_200000-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 888.3 ± 5.7 | 871.2 | 905.9 | 1.09 ± 0.02 |
| `base-bigint` | 1119.7 ± 8.6 | 1106.2 | 1139.5 | 1.38 ± 0.02 |
| `ranges-felt` | 813.7 ± 11.0 | 782.0 | 838.6 | 1.00 |
| `ranges-bigint` | 1059.3 ± 33.7 | 1039.6 | 1281.1 | 1.30 ± 0.05 |

dict_integration_benchmark-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.127 ± 0.008 | 1.115 | 1.145 | 1.07 ± 0.01 |
| `base-bigint` | 1.327 ± 0.006 | 1.318 | 1.357 | 1.26 ± 0.01 |
| `ranges-felt` | 1.054 ± 0.004 | 1.049 | 1.072 | 1.00 |
| `ranges-bigint` | 1.252 ± 0.007 | 1.234 | 1.273 | 1.19 ± 0.01 |

dict_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 741.6 ± 7.1 | 732.5 | 768.3 | 1.08 ± 0.02 |
| `base-bigint` | 922.8 ± 7.6 | 913.0 | 938.7 | 1.35 ± 0.03 |
| `ranges-felt` | 685.3 ± 11.6 | 667.7 | 744.6 | 1.00 |
| `ranges-bigint` | 874.7 ± 5.5 | 868.7 | 903.6 | 1.28 ± 0.02 |

ec_op-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 4.725 ± 0.080 | 4.668 | 5.117 | 1.01 ± 0.03 |
| `base-bigint` | 8.449 ± 0.118 | 8.336 | 8.897 | 1.81 ± 0.05 |
| `ranges-felt` | 4.674 ± 0.104 | 4.489 | 5.229 | 1.00 |
| `ranges-bigint` | 8.510 ± 0.101 | 8.343 | 8.758 | 1.82 ± 0.05 |

ec_op-quick.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 4.615 ± 0.047 | 4.561 | 4.784 | 1.01 ± 0.02 |
| `base-bigint` | 8.433 ± 0.100 | 8.197 | 8.761 | 1.84 ± 0.04 |
| `ranges-felt` | 4.578 ± 0.069 | 4.475 | 4.872 | 1.00 |
| `ranges-bigint` | 8.380 ± 0.087 | 8.297 | 8.682 | 1.83 ± 0.03 |

factorial_multirun-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.548 ± 0.017 | 1.518 | 1.601 | 1.00 |
| `base-bigint` | 1.938 ± 0.028 | 1.912 | 2.056 | 1.25 ± 0.02 |
| `ranges-felt` | 1.561 ± 0.006 | 1.551 | 1.576 | 1.01 ± 0.01 |
| `ranges-bigint` | 1.979 ± 0.014 | 1.937 | 2.019 | 1.28 ± 0.02 |

factorial_multirun-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 854.1 ± 13.3 | 822.4 | 885.6 | 1.00 |
| `base-bigint` | 1259.2 ± 23.4 | 1218.8 | 1337.2 | 1.47 ± 0.04 |
| `ranges-felt` | 856.1 ± 9.6 | 845.2 | 879.5 | 1.00 ± 0.02 |
| `ranges-bigint` | 1292.8 ± 25.0 | 1252.7 | 1374.4 | 1.51 ± 0.04 |

fibonacci_1000_multirun-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.592 ± 0.012 | 1.557 | 1.609 | 1.00 |
| `base-bigint` | 1.729 ± 0.010 | 1.716 | 1.765 | 1.09 ± 0.01 |
| `ranges-felt` | 1.597 ± 0.010 | 1.557 | 1.612 | 1.00 ± 0.01 |
| `ranges-bigint` | 1.762 ± 0.013 | 1.721 | 1.797 | 1.11 ± 0.01 |

fibonacci_1000_multirun-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 809.8 ± 8.3 | 798.5 | 843.2 | 1.00 |
| `base-bigint` | 981.9 ± 10.3 | 970.1 | 1017.9 | 1.21 ± 0.02 |
| `ranges-felt` | 824.7 ± 9.9 | 810.5 | 867.6 | 1.02 ± 0.02 |
| `ranges-bigint` | 1026.5 ± 6.9 | 1009.4 | 1043.0 | 1.27 ± 0.02 |

field_arithmetic_get_square_benchmark-proof.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 73.7 ± 1.8 | 72.4 | 85.8 | 1.05 ± 0.03 |
| `base-bigint` | 74.8 ± 0.4 | 74.4 | 77.1 | 1.07 ± 0.01 |
| `ranges-felt` | 69.9 ± 0.4 | 69.3 | 71.3 | 1.00 |
| `ranges-bigint` | 70.0 ± 0.3 | 69.6 | 70.7 | 1.00 ± 0.01 |

field_arithmetic_get_square_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 42.7 ± 0.2 | 42.3 | 43.3 | 1.07 ± 0.01 |
| `base-bigint` | 45.9 ± 0.2 | 45.4 | 47.0 | 1.15 ± 0.02 |
| `ranges-felt` | 40.0 ± 0.5 | 39.5 | 41.9 | 1.00 |
| `ranges-bigint` | 42.9 ± 0.5 | 42.2 | 44.5 | 1.07 ± 0.02 |

find_element-proof.md

find_element-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 507.4 ± 4.7 | 498.8 | 527.2 | 1.11 ± 0.02 |
| `base-bigint` | 647.7 ± 6.4 | 638.2 | 666.7 | 1.42 ± 0.03 |
| `ranges-felt` | 455.2 ± 7.8 | 432.8 | 473.7 | 1.00 |
| `ranges-bigint` | 623.6 ± 12.7 | 591.3 | 690.8 | 1.37 ± 0.04 |

integration_builtins-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 5.650 ± 0.104 | 5.512 | 6.060 | 1.13 ± 0.02 |
| `base-bigint` | 4.999 ± 0.050 | 4.938 | 5.084 | 1.00 |
| `ranges-felt` | 5.540 ± 0.049 | 5.409 | 5.676 | 1.11 ± 0.01 |
| `ranges-bigint` | 5.134 ± 0.065 | 5.080 | 5.395 | 1.03 ± 0.02 |

integration_builtins-quick.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.020 ± 0.004 | 1.016 | 1.031 | 1.00 |
| `base-bigint` | 1.104 ± 0.006 | 1.091 | 1.116 | 1.08 ± 0.01 |
| `ranges-felt` | 1.030 ± 0.006 | 1.022 | 1.060 | 1.01 ± 0.01 |
| `ranges-bigint` | 1.131 ± 0.011 | 1.118 | 1.176 | 1.11 ± 0.01 |

keccak_integration_benchmark-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 6.150 ± 0.134 | 5.937 | 6.564 | 1.21 ± 0.03 |
| `base-bigint` | 5.568 ± 0.108 | 5.392 | 6.036 | 1.09 ± 0.02 |
| `ranges-felt` | 5.575 ± 0.041 | 5.523 | 5.722 | 1.09 ± 0.01 |
| `ranges-bigint` | 5.099 ± 0.053 | 5.059 | 5.324 | 1.00 |

keccak_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 984.8 ± 20.9 | 956.0 | 1021.2 | 1.06 ± 0.03 |
| `base-bigint` | 1085.9 ± 17.0 | 1052.5 | 1121.2 | 1.17 ± 0.03 |
| `ranges-felt` | 929.2 ± 15.0 | 905.9 | 959.7 | 1.00 |
| `ranges-bigint` | 1059.2 ± 16.9 | 1030.7 | 1091.3 | 1.14 ± 0.03 |

linear_search-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.783 ± 0.015 | 1.761 | 1.835 | 1.08 ± 0.01 |
| `base-bigint` | 1.980 ± 0.011 | 1.960 | 2.000 | 1.20 ± 0.01 |
| `ranges-felt` | 1.651 ± 0.015 | 1.612 | 1.714 | 1.00 |
| `ranges-bigint` | 1.846 ± 0.009 | 1.826 | 1.878 | 1.12 ± 0.01 |

linear_search-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 931.2 ± 7.6 | 904.1 | 965.9 | 1.10 ± 0.01 |
| `base-bigint` | 1159.0 ± 8.1 | 1149.2 | 1196.3 | 1.37 ± 0.02 |
| `ranges-felt` | 845.5 ± 9.1 | 823.3 | 887.2 | 1.00 |
| `ranges-bigint` | 1077.1 ± 11.9 | 1066.6 | 1126.4 | 1.27 ± 0.02 |

math_cmp_and_pow_integration_benchmark-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.148 ± 0.005 | 1.138 | 1.159 | 1.06 ± 0.01 |
| `base-bigint` | 1.331 ± 0.004 | 1.324 | 1.342 | 1.23 ± 0.01 |
| `ranges-felt` | 1.080 ± 0.005 | 1.071 | 1.104 | 1.00 |
| `ranges-bigint` | 1.258 ± 0.015 | 1.246 | 1.351 | 1.17 ± 0.01 |

math_cmp_and_pow_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 765.1 ± 3.9 | 757.4 | 775.6 | 1.09 ± 0.01 |
| `base-bigint` | 950.7 ± 6.8 | 942.4 | 975.6 | 1.36 ± 0.01 |
| `ranges-felt` | 700.4 ± 5.1 | 694.0 | 723.4 | 1.00 |
| `ranges-bigint` | 889.0 ± 7.9 | 879.7 | 931.9 | 1.27 ± 0.01 |

math_integration_benchmark-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.115 ± 0.008 | 1.103 | 1.145 | 1.07 ± 0.01 |
| `base-bigint` | 1.259 ± 0.008 | 1.245 | 1.295 | 1.21 ± 0.01 |
| `ranges-felt` | 1.043 ± 0.005 | 1.038 | 1.071 | 1.00 |
| `ranges-bigint` | 1.178 ± 0.004 | 1.169 | 1.185 | 1.13 ± 0.01 |

math_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 751.3 ± 7.5 | 745.1 | 775.4 | 1.12 ± 0.01 |
| `base-bigint` | 897.7 ± 6.3 | 888.7 | 923.6 | 1.34 ± 0.01 |
| `ranges-felt` | 671.8 ± 3.3 | 667.4 | 680.8 | 1.00 |
| `ranges-bigint` | 832.3 ± 9.8 | 817.8 | 876.7 | 1.24 ± 0.02 |

memory_integration_benchmark-proof.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 935.2 ± 12.3 | 915.0 | 970.0 | 1.07 ± 0.02 |
| `base-bigint` | 1072.4 ± 12.8 | 1048.5 | 1129.3 | 1.22 ± 0.02 |
| `ranges-felt` | 875.8 ± 5.9 | 868.6 | 898.0 | 1.00 |
| `ranges-bigint` | 1005.1 ± 10.0 | 992.9 | 1056.5 | 1.15 ± 0.01 |

memory_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 594.1 ± 11.4 | 574.5 | 630.0 | 1.11 ± 0.03 |
| `base-bigint` | 741.0 ± 11.7 | 729.7 | 799.7 | 1.39 ± 0.03 |
| `ranges-felt` | 533.2 ± 9.3 | 516.5 | 571.4 | 1.00 |
| `ranges-bigint` | 683.3 ± 13.7 | 672.4 | 735.6 | 1.28 ± 0.03 |

operations_with_data_structures_benchmarks-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.121 ± 0.005 | 1.112 | 1.131 | 1.08 ± 0.01 |
| `base-bigint` | 1.238 ± 0.008 | 1.229 | 1.278 | 1.20 ± 0.01 |
| `ranges-felt` | 1.036 ± 0.003 | 1.031 | 1.043 | 1.00 |
| `ranges-bigint` | 1.135 ± 0.004 | 1.128 | 1.146 | 1.10 ± 0.00 |

operations_with_data_structures_benchmarks-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 672.2 ± 7.6 | 662.0 | 689.4 | 1.11 ± 0.01 |
| `base-bigint` | 781.3 ± 5.3 | 772.8 | 804.1 | 1.29 ± 0.01 |
| `ranges-felt` | 606.3 ± 4.4 | 595.0 | 619.6 | 1.00 |
| `ranges-bigint` | 733.8 ± 5.5 | 726.1 | 753.0 | 1.21 ± 0.01 |

pedersen-proof.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 407.3 ± 5.1 | 399.2 | 424.7 | 1.00 |
| `base-bigint` | 427.3 ± 4.0 | 417.4 | 435.8 | 1.05 ± 0.02 |
| `ranges-felt` | 411.7 ± 6.7 | 399.1 | 431.1 | 1.01 ± 0.02 |
| `ranges-bigint` | 429.4 ± 7.4 | 418.3 | 455.6 | 1.05 ± 0.02 |

pedersen-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 320.4 ± 3.7 | 306.7 | 331.7 | 1.00 ± 0.02 |
| `base-bigint` | 335.5 ± 7.2 | 325.7 | 350.1 | 1.05 ± 0.03 |
| `ranges-felt` | 320.3 ± 5.7 | 308.9 | 329.6 | 1.00 |
| `ranges-bigint` | 339.0 ± 10.0 | 327.3 | 374.3 | 1.06 ± 0.04 |

poseidon_integration_benchmark-proof.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 800.0 ± 3.0 | 796.1 | 815.7 | 1.03 ± 0.01 |
| `base-bigint` | 846.5 ± 3.4 | 843.5 | 867.1 | 1.09 ± 0.01 |
| `ranges-felt` | 773.3 ± 6.6 | 758.0 | 784.0 | 1.00 |
| `ranges-bigint` | 840.2 ± 5.3 | 834.6 | 866.6 | 1.09 ± 0.01 |

poseidon_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 715.4 ± 2.5 | 712.2 | 722.3 | 1.03 ± 0.00 |
| `base-bigint` | 771.3 ± 1.1 | 769.0 | 773.2 | 1.11 ± 0.00 |
| `ranges-felt` | 693.9 ± 2.3 | 689.6 | 702.5 | 1.00 |
| `ranges-bigint` | 755.7 ± 1.9 | 751.5 | 762.0 | 1.09 ± 0.00 |

search_sorted_lower-proof.md

search_sorted_lower-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 313.5 ± 7.6 | 298.9 | 335.1 | 1.12 ± 0.03 |
| `base-bigint` | 375.1 ± 5.4 | 364.9 | 386.0 | 1.34 ± 0.03 |
| `ranges-felt` | 279.8 ± 4.8 | 269.2 | 292.2 | 1.00 |
| `ranges-bigint` | 349.9 ± 2.5 | 346.0 | 358.3 | 1.25 ± 0.02 |

secp_integration_benchmark-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.347 ± 0.013 | 1.322 | 1.398 | 1.05 ± 0.01 |
| `base-bigint` | 1.511 ± 0.014 | 1.501 | 1.577 | 1.17 ± 0.01 |
| `ranges-felt` | 1.286 ± 0.011 | 1.261 | 1.335 | 1.00 |
| `ranges-bigint` | 1.431 ± 0.005 | 1.423 | 1.449 | 1.11 ± 0.01 |

secp_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 846.5 ± 10.1 | 824.5 | 870.1 | 1.03 ± 0.01 |
| `base-bigint` | 1028.2 ± 8.2 | 1003.1 | 1044.2 | 1.25 ± 0.01 |
| `ranges-felt` | 821.2 ± 5.2 | 815.5 | 842.7 | 1.00 |
| `ranges-bigint` | 1004.2 ± 4.4 | 997.7 | 1018.0 | 1.22 ± 0.01 |

set_integration_benchmark-proof.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 560.9 ± 7.8 | 544.1 | 579.9 | 1.03 ± 0.02 |
| `base-bigint` | 849.5 ± 85.0 | 811.8 | 1424.5 | 1.56 ± 0.16 |
| `ranges-felt` | 544.3 ± 9.6 | 528.3 | 567.8 | 1.00 |
| `ranges-bigint` | 827.3 ± 20.5 | 800.5 | 885.4 | 1.52 ± 0.05 |

set_integration_benchmark-quick.md
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 452.1 ± 7.3 | 438.6 | 468.2 | 1.03 ± 0.02 |
| `base-bigint` | 738.9 ± 17.6 | 714.0 | 788.8 | 1.68 ± 0.04 |
| `ranges-felt` | 440.5 ± 5.1 | 427.7 | 450.6 | 1.00 |
| `ranges-bigint` | 727.5 ± 8.5 | 704.3 | 756.4 | 1.65 ± 0.03 |

uint256_integration_benchmark-proof.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 3.527 ± 0.041 | 3.453 | 3.603 | 1.11 ± 0.01 |
| `base-bigint` | 3.470 ± 0.041 | 3.434 | 3.683 | 1.09 ± 0.01 |
| `ranges-felt` | 3.266 ± 0.031 | 3.197 | 3.375 | 1.03 ± 0.01 |
| `ranges-bigint` | 3.176 ± 0.017 | 3.092 | 3.237 | 1.00 |

uint256_integration_benchmark-quick.md
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `base-felt` | 1.150 ± 0.012 | 1.102 | 1.179 | 1.05 ± 0.01 |
| `base-bigint` | 1.361 ± 0.021 | 1.302 | 1.391 | 1.25 ± 0.02 |
| `ranges-felt` | 1.091 ± 0.009 | 1.080 | 1.129 | 1.00 |
| `ranges-bigint` | 1.277 ± 0.008 | 1.257 | 1.315 | 1.17 ± 0.01 |

### CODE REVIEW

File: vm/src/types/program.rs [hunk 1]
```diff
@@ -83,12 +85,39 @@ impl Program {
...
+                *s += v;
+                Some(res)
+            });
+
+        let mut hints_ranges = vec![None; hints.last().map(|(x, _)| x + 1).unwrap_or(0)];
+        for (pc, r) in hints_ranges_iter {
+            hints_ranges[*pc] = Some(r);
+        }
+
+        let hints = hints.into_iter().flat_map(|(_, v)| v).collect();
```
Comment (MegaRedHand) [inline vm/src/types/program.rs:112]: I think we should extract this part into a helper function

File: vm/src/serde/deserialize_program.rs [hunk 2]
```diff
@@ -418,9 +419,41 @@ pub fn parse_program_json(
...
+            );
+            *s += v;
+            Some(res)
+        });
+
+    for (pc, r) in hints_ranges_iter {
+        hints_ranges[*pc] = Some(r);
+    }
+
+    let hints = program_hints.into_iter().flat_map(|(_, v)| v).collect();
```
Comment (MegaRedHand) [inline vm/src/serde/deserialize_program.rs:451]: Here's the same bit of code

File: vm/src/utils.rs [hunk 3]
```diff
@@ -342,10 +344,42 @@ pub mod test_utils {
...
+                    );
+                    *s += v;
+                    Some(res)
+                });
+
+            for (pc, r) in hints_ranges_iter {
+                hints_ranges[*pc] = Some(r);
+            }
+
+            let hints = hints.into_iter().flat_map(|(_, v)| v).collect();
```
Comment (MegaRedHand) [inline vm/src/utils.rs:376]: And here again

File: vm/src/types/program.rs (hunk 1 above)
Comment (Oppen) [reply vm/src/types/program.rs:112]: Hehehe I had a TODO I removed because I considered it for a later refactor.
Will do now then.

**Review by MegaRedHand — APPROVED** [review]

**Review by fmoletta — APPROVED** [review]
Nice!

File: vm/src/serde/deserialize_program.rs [hunk 4]
```diff
@@ -812,43 +816,60 @@ mod tests {
...
         assert_eq!(program.shared_program_data.data, data);
         assert_eq!(program.shared_program_data.main, Some(0));
-        assert_eq!(program.shared_program_data.hints, hints);
+
+        let program_hints: HashMap<_, _> = program
+            .shared_program_data
+            .hints_ranges
+            .iter()
+            .enumerate()
+            .filter_map(|(pc, r)| r.map(|(s, l)| (pc, (s, s + l.get()))))
```
Comment (mmsc2) [inline vm/src/serde/deserialize_program.rs:869]: this is a bit confusing, maybe a comment explaining what is happening would be useful.

File: vm/src/serde/deserialize_program.rs [hunk 5]
```diff
@@ -871,43 +892,57 @@ mod tests {
...
         assert_eq!(program.shared_program_data.data, data);
         assert_eq!(program.shared_program_data.main, None);
-        assert_eq!(program.shared_program_data.hints, hints);
+
+        let program_hints: HashMap<_, _> = program
+            .shared_program_data
+            .hints_ranges
+            .iter()
+            .enumerate()
+            .filter_map(|(pc, r)| r.map(|(s, l)| (pc, (s, s + l.get()))))
```
Comment (mmsc2) [inline vm/src/serde/deserialize_program.rs:942]: same as above

**Review by mmsc2 — APPROVED** [review]

File: vm/src/serde/deserialize_program.rs (hunk 4 above)
Comment (Oppen) [reply vm/src/serde/deserialize_program.rs:869]: Thanks, I'll do that, or maybe I'll see if I can make it simpler.

**Pull request merged**